### PR TITLE
GPU refactor common code from ECAL multifit and HCAL mahi

### PR DESCRIPTION
**Pull request by mariadalfonso**
move here function used both in 
RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu 
and 
RecoLocalCalo/HcalRecProducers/src/MahiGPU.cu

@fwyzard

### CODE REVIEW

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@mariadalfonso could you change the title of the PR to be a bit more descriptive ?
For example
> Refactor common code from ECAL multifit and HCAL mahi

@jpata @perrotta I can rebase #31719 and #31720 and remove the common code once this is merged.

**Comment by davidlange6** [comment]
Hi @mariadalfonso - dependency-wise, what drives this to go to DataFormats? (its clearly not a data format but an algorithm). Could it instead go to a package like RecoLocalCalo/CommonAlgos (would be new) or CommonTools/RecoAlgos (if the usage is to grow beyond RecoLocalCalo)?

**Comment by fwyzard** [comment]
Hi @davidlange6, the use of `DataFormats/CaloRecHit` was introduced by https://github.com/cms-sw/cmssw/pull/31704 becase `DataFormats/Math` was deemed too generic for calorimeter code.

If necessary, things can be moved again of course, but I would agree to that only after all PRs depending on it have been merged.

**Comment by jpata** [comment]
> Could it instead go to a package like RecoLocalCalo/CommonAlgos (would be new) or CommonTools/RecoAlgos (if the usage is to grow beyond RecoLocalCalo)?
> If necessary, things can be moved again of course, but I would agree to that only after all PRs depending on it have been merged.

The minimal suggestion to move this from DataFormats/Math -> DataFormats/CaloRecHit was to discourage this apparently calorimeter-specific code to be relied on by others for generic math operations. I agree that as it's not a dataformat as such, it would have a better home in some non-DataFormats package - thx for the suggestion.

> I can rebase #31719 and #31720 and remove the common code once this is merged.

Sounds good to me.

**Comment by slava77** [comment]
> Hi @mariadalfonso - dependency-wise, what drives this to go to DataFormats? (its clearly not a data format but an algorithm). Could it instead go to a package like RecoLocalCalo/CommonAlgos (would be new) or CommonTools/RecoAlgos (if the usage is to grow beyond RecoLocalCalo)?

I agree that the algorithms should be moved out of DataFormats.

Still, we tend to keep compiled matrix and basic methods for them in the DataFormats.
Earlier on, IIRC, the use cases  started as support for straightforward (although limited in variety of sizes) matrix operations for Eigen, which did fit fine in DataFromats/Math. 
Also, historically we put some algorithmic code in DataFormats if it needs/can be available via FWLite.
Is this the case here, can this Eigen code be executed in FWLite?

File: DataFormats/CaloRecHit/interface/MultifitComputations.h [hunk 1]
```diff
@@ -283,6 +283,94 @@ namespace calo {
...
+          accum[counter] = -inputAmplitudesView(counter);
+
+        // iterate
+        for (int icol = 0; icol < NPULSES; icol++) {
+          float pm_col[NSAMPLES];
+
+          // preload a column of pulse matrix
+#pragma unroll
+          for (int counter = 0; counter < NSAMPLES; counter++)
+            pm_col[counter] = __ldg(&pulseMatrixView.coeffRef(counter, icol));
```
Comment (mariadalfonso) [inline DataFormats/CaloRecHit/interface/MultifitComputations.h:320]: @fwyzard 
do I need to add here 
#ifdef __CUDA_ARCH__
as you did in 
https://github.com/cms-sw/cmssw/blob/5bef34a0f16bb3b81f3c7306994550a22db30fba/DataFormats/CaloRecHit/interface/MultifitComputations.h#L210
?

Comment (fwyzard) [reply DataFormats/CaloRecHit/interface/MultifitComputations.h:320]: Ah, yes right.

`__ldg` is an intrinsic understood only by CUDA.

Comment (fwyzard) [inline DataFormats/CaloRecHit/interface/MultifitComputations.h:320]:
```suggestion
#ifdef __CUDA_ARCH__ 
            pm_col[counter] = __ldg(&pulseMatrixView.coeffRef(counter, icol));
#else
            pm_col[counter] = pulseMatrixView.coeffRef(counter, icol);
#endif
```

**Comment by jpata** [comment]
enable gpu

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Per the discussion in https://github.com/cms-sw/cmssw/issues/31735, there is some motivation to reconsider the use of Eigen dependencies in DataFormats, and thus moving the common calo methods as David has suggested. I would not set this as a blocker for this PR necessarily (considering the timeline of the patatrack integration effort), but to be kept in mind as an additional data point.

Although no new unit tests were added for the new functions in this PR (see https://github.com/cms-sw/cmssw/issues/31880), it would be good to make sure the existing unit tests at least are running.

It seems like with `enable gpu`, the CUDA unit tests are not being run on the jenkins infrastructure. @smuzaffar could you comment if this would be possible?

I was trying instead to run the test on my local GPU machine as I have done successfully in the past for the previous patatrack PR https://github.com/cms-sw/cmssw/pull/31704#issuecomment-712985097, however, the latest cuda update in CMSSW https://github.com/cms-sw/cmsdist/commit/e4129585e82760dadb457b88e4254045497b8da1 requires driver updates which I'm addressing on my infrastructure.

**Comment by smuzaffar** [comment]
@jpata, Yes this is on my to do list but will take some time (hopefully by the end of the year)

**Comment by jpata** [comment]
> Yes this is on my to do list but will take some time (hopefully by the end of the year)

OK, thank you for confirming (also that I didn't misunderstand whether/how this is supposed to work)!

Meanwhile, I was able to run the existing unit test successfully at least on my machine:
```bash
$ ./test/slc7_amd64_gcc820/test_calo_rechit 
0, 10 GeV, 1 ns  flags=0x0  aux=0x0 
0, 10 GeV, 1 ns  flags=0x0  aux=0x0 
all good!
```

I have no further comments from the reco side, modulo the Eigen-in-Dataformats / package location issue.

**Comment by mariadalfonso** [comment]
> @fwyzard  wrote 
> If necessary, things can be moved again of course, but I would agree to that only after all PRs depending on it have been merged.

@jpata 
is this fine ? 
When is Ecal/Hcal is all integrated/reviewed it will be more clear what is pure Eigen algebra and caloMultifit algorithm and how to keep in synch c++/gpu code and what is cuda specific or more general

**Comment by fwyzard** [comment]
> I have no further comments from the reco side, modulo the Eigen-in-Dataformats / package location issue.

To keep track of this and other requests, I would suggest to create an issue on GitHub to collect all the changes that should be made after the initial PRs have been merged.

**Comment by jpata** [comment]
I opened an issue for this https://github.com/cms-sw/cmssw/issues/32159, from my side, I would be OK to move forward like this. I'll wait for comments before proceeding with the reco signature.

**Comment by fwyzard** [comment]
@jpata, could you sign this for RECO to have it merged and allow testing more properly the ECAL and HCAL PRs ?

**Comment by jpata** [comment]
+reconstruction

- adds common code to DataFormats/CaloRecHit needed for patatrack PRs #31719, #31720
- existing tests pass as expected (new code is not included in tests)
- possibly this package should be moved later, see #32159
- it was pointed out that this code contains a number of code rule violations already from #31704 (single character variables, uppercase variable names), I added this to the issue.

**Comment by silviodonato** [comment]
+1

**Pull request merged**